From 47bf5af50331bb81a6bc09b3859ebcc52e98e10a Mon Sep 17 00:00:00 2001
From: Yuvasree Sivasankaran <quic_ysivasan@quicinc.com>
Date: Mon, 24 Jun 2024 22:37:57 +0530
Subject: [PATCH] QSDK: wifi: ath12k: Remove unneccesary checks in DL fast path

Remove msdu destructor and skb frag list check in fast
tx completion since the checks are invalid in fast path.
And unneccesary len initialization in fast tx path.

Signed-off-by: Yuvasree Sivasankaran <quic_ysivasan@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 20 +++-----------------
 1 file changed, 3 insertions(+), 17 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -281,11 +281,6 @@ int ath12k_dp_tx_direct(struct ath12k_li
 #endif
 	dma_addr_t paddr;
 	int ret;
-#ifdef CPTCFG_MAC80211_SFE_SUPPORT
-	int len = 256;
-#else
-	int len = skb->len;
-#endif
 
 	if (unlikely(test_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags)))
 		return -ESHUTDOWN;
@@ -299,7 +294,7 @@ int ath12k_dp_tx_direct(struct ath12k_li
 		return -ENOSPC;
 	}
 
-	dmac_clean_range_no_dsb(skb->data, skb->data + len);
+	dmac_clean_range_no_dsb(skb->data, skb->data + DP_TX_SFE_BUFFER_SIZE);
 	paddr = (dma_addr_t)virt_to_phys(skb->data);
 
 	tx_desc->skb = skb;
@@ -838,14 +833,6 @@ ath12k_dp_tx_htt_tx_complete_buf(struct
 
 	/* Free skb here if stats is disabled */
 	if (ab->stats_disable && tx_desc->flags & DP_TX_DESC_FLAG_FAST) {
-		if (msdu->destructor) {
-			msdu->wifi_acked_valid = 1;
-			msdu->wifi_acked = htt_status;
-		}
-		if (skb_has_frag_list(msdu)) {
-			kfree_skb_list(skb_shinfo(msdu)->frag_list);
-			skb_shinfo(msdu)->frag_list = NULL;
-		}
 		dev_kfree_skb(msdu);
 		return;
 	}
@@ -1668,18 +1655,9 @@ int ath12k_dp_tx_completion_handler(stru
 		ab->soc_stats.tx_completed[data_ring_id]++;
 
 		/* Free skb here if stats is disabled */
-		if (ab->stats_disable && tx_desc->flags & DP_TX_DESC_FLAG_FAST) {
-			if (unlikely(msdu->destructor)) {
-				msdu->wifi_acked_valid = 1;
-				msdu->wifi_acked = htt_status == HAL_WBM_TQM_REL_REASON_FRAME_ACKED;
-			}
-			if (unlikely(skb_has_frag_list(msdu))) {
-				kfree_skb_list(skb_shinfo(msdu)->frag_list);
-				skb_shinfo(msdu)->frag_list = NULL;
-			}
+		if (ab->stats_disable && tx_desc->flags & DP_TX_DESC_FLAG_FAST)
 			__skb_queue_head(&free_list_head, msdu);
-
-		} else {
+		else {
 
 			ath12k_dp_tx_complete_msdu(ar, msdu, tx_status, buf_rel_source,
 						   tx_desc, timestamp, htt_status);
--- a/drivers/net/wireless/ath/ath12k/dp.h
+++ b/drivers/net/wireless/ath/ath12k/dp.h
@@ -197,6 +197,7 @@ struct ath12k_pdev_dp {
 #define DP_RX_MONITOR_BUF_LOW_TH	32
 #define DP_TX_MONITOR_BUF_RING_SIZE	4096
 #define DP_TX_MONITOR_DEST_RING_SIZE	2048
+#define DP_TX_SFE_BUFFER_SIZE           256
 
 #define DP_TX_MONITOR_BUF_SIZE		2048
 #define DP_TX_MONITOR_BUF_SIZE_MIN	48
